### PR TITLE
correct versions order and show only indexed versions

### PR DESCRIPTION
**Pull request by sonalkr132**
Looks like I have found more tests which fail only on travis :sweat: 

[preserve order test on bunder-api](https://github.com/bundler/bundler-api/blob/master/spec/gem_info_spec.rb#L64-L68).
[Indexed test on bundler-api](https://github.com/bundler/bundler-api/blob/master/spec/gem_info_spec.rb#L70-L95).

### CODE REVIEW

**Comment by sonalkr132** [comment]
[Build on my travis](https://travis-ci.org/sonalkr132/rubygems.org/jobs/132070056) for 33c6db2

Only test which fail on both build don't have same failure message either:
**Build here:**
```
GemDependentTest#test_: with gem_names no dependencies should return all versions for a gem.  [/home/travis/build/rubygems/rubygems.org/test/unit/gem_dependent_test.rb:33]:
--- expected
+++ actual
@@ -1 +1 @@
-[{:name=>"rack", :number=>"0.0.1", :platform=>"ruby", :dependencies=>[]}, {:name=>"rack", :number=>"0.0.2", :platform=>"ruby", :dependencies=>[]}]
+[{:name=>"rack", :number=>"0.0.1", :platform=>"ruby", :dependencies=>[]}]
```
**My build:**
```
GemDependentTest#test_: with gem_names non indexed versions should filter non indexed version.  [/home/travis/build/sonalkr132/rubygems.org/test/unit/gem_dependent_test.rb:78]:
--- expected
+++ actual
@@ -1 +1 @@
-[{:name=>"rack", :number=>"0.0.1", :platform=>"ruby", :dependencies=>[]}]
+[{:name=>"rack", :number=>"0.0.1", :platform=>"ruby", :dependencies=>[["foo", ">= 0.0.0"]]}, {:name=>"rack", :number=>"0.2.0", :platform=>"ruby", :dependencies=>[]}, {:name=>"rack", :number=>"1.0.1", :platform=>"ruby", :dependencies=>[]}]
```

**Comment by sonalkr132** [comment]
Although test have passed and they are correct, I think we should figure out why they were failing previously.
I had to use different `gem names` because database is not cleaning between test runs (only on travis). Objects created in unit tests were breaking functional test just cause they used the same name ([travis link](https://travis-ci.org/rubygems/rubygems.org/jobs/132164054#L923)).

It is exactly what was making #1281 fail ([travis build link](https://travis-ci.org/rubygems/rubygems.org/jobs/131903004#L918)).

This is all I know so far and I don't know where to go with this.

**Comment by dwradcliffe** [comment]
The code looks good. I'm not sure about the Travis failures, and we should figure that out, but I think we should proceed with this anyway.

**Comment by sonalkr132** [comment]
Here, we are ordering just by `number` whereas [bundler-api orderes versions ](https://github.com/bundler/bundler-api/blob/master/lib/bundler_api/gem_info.rb#L31) with `created_at`, `number`, `platform`, and dependencies by `name`.
Do we need all of that? Do we know someone who can answer it?

@fotanus git blame says its you :crystal_ball:

**Comment by dwradcliffe** [comment]
@indirect @segiddins might know

**Comment by fotanus** [comment]
I'm in a run right now, but the orderings should be consistent across both bundler-api and rubygems.org. If is not the case, it is wrong.

**Comment by arthurnn** [comment]
Regarding data leaking between tests. We should be using transactional fixtures , which should not let data leaky. Can you check if we have a test that disable it, and dont clean the data afterwards?

**Comment by arthurnn** [comment]
> Here, we are ordering just by number whereas bundler-api orderes versions with created_at, number, platform, and dependencies by name (I hope I got this right!).
Do we need all of that? 

@sonalkr132 I think it is a good idea to maintain that order.

**Comment by fotanus** [comment]
Now with a bit more of time, I think that we should keep the most restrictive order possible, trying to avoid a database update to break the order and cause trouble. So I suggest copy the order from bundler-api. I have something pending already for bundler, so if you guys are in rush feel free to fix this!

**Comment by sonalkr132** [comment]
>  I suggest copy the order from bundler-api.

I am getting more and more inclined towards using raw sql in `fetch_dependency_from_db`. That way we can almost copy [the exact query](https://github.com/bundler/bundler-api/blob/master/lib/bundler_api/gem_info.rb#L16-L31) from bundler-api. I also ran performance test on dependency end point and `fetch_dependency_from_db` took 27% of request time with one record in DB(10.96% of 40% request time. 60% was spent in `setup`). 

screenshot from 2016-05-24 17-14-25

report: https://jsfiddle.net/pudg03g7/

**Comment by fotanus** [comment]
I think @dwradcliffe can help with the code style. I'm good either way.

**Comment by arthurnn** [comment]
@sonalkr132 I am totally for it to use raw SQL here.. As long as we sanitize the parameters for security reasons, thats what we should do IMO.

**Comment by sonalkr132** [comment]
I have updated it to the exact code of bundler-api, except I have dropped [ordering of nested SELECT query](https://github.com/bundler/bundler-api/blob/master/lib/bundler_api/gem_info.rb#L25). We were ordering again at end so nested ordering didn't make any difference other than slowing the query.

This has one more difference, angle brackets `>` used in requirements are rendered as `["rake","\u003e= 0.8.3"]`. It is not introduced in this PR, rather it was always there. I suppose thats rails default escaping. This [stackoverflow answer](http://stackoverflow.com/questions/5123993/json-encoding-wrongly-escaped-rails-3-ruby-1-9-2#answer-5124452) suggests that it won't make a difference:

> The various gsub calls are forcing non-ASCII UTF-8 to the \uXXXX notation that you're seeing. Hex encoded UTF-8 should be acceptable to anything that processes JSON but you could always post-process the JSON (or monkey patch in a modified JSON escaper) to convert the \uXXXX notation to raw UTF-8 if necessary.

File: test/functional/api/v1/dependencies_controller_test.rb [hunk 1]
```diff
@@ -78,23 +78,23 @@ class Api::V1::DependenciesControllerTest < ActionController::TestCase
...
         {
           'name'              => 'myrails',
-          'number'            => '3.0.0',
+          'number'            => '1.0.0',
           'platform'          => 'ruby',
           'rubygems_version'  => '>= 2.6.3',
           'ruby_version'      => '>= 2.0.0',
           'checksum'          => 'tdQEXD9Gb6kf4sxqvnkjKhpXzfEE96JucW4KHieJ33g=',
-          'created_at'        => '2016-05-24T00:00:00.000Z',
+          'created_at'        => '2016-05-24 00:00:00',
```
Comment (dwradcliffe) [inline test/functional/api/v1/dependencies_controller_test.rb]: This is closer but still not exactly what the current format is.
Currently I'm seeing: `2016-05-24 00:00:00 +0000`

Comment (sonalkr132) [inline test/functional/api/v1/dependencies_controller_test.rb]: :+1:

**Pull request merged**